src/viam/sdk/tests/test_utils.hpp

```diff
@@ -12,5 +16,61 @@ using namespace viam::sdk;
 AttributeMap fake_map();
 std::vector<GeometryConfig> fake_geometries();
 
+// TestServer is a wrapper around viam::sdk::Server that is a friend of the
+// class and can therefore access its private fields.
+class TestServer {
+   public:
+    TestServer(std::shared_ptr<Server> sdk_server);
+    ~TestServer();
+
+    std::shared_ptr<grpc::Channel> grpc_in_process_channel(const grpc::ChannelArguments& args);
+
+   private:
+    std::shared_ptr<Server> sdk_server_;
+};
+
+// This function sets up the following architecture:
+// -- ResourceClient
+//        /\
+//
+//        | (grpc InProcessChannel)
+//
+//        \/
+// -- ResourceServer
+//        /\
+//
+//        | (function calls)
+//
+//        \/
+// -- MockResource
+//
+// This is as close to a real client->server->resource setup as we can get
+// without starting another process.
+//
+// The passed in test_case function will have access to the created ResourceClient.
+template <typename ClientType, typename ServerType, typename F>
+void client_to_mock_pipeline(std::shared_ptr<Resource> mock, F&& test_case) {
+    // Create a viam RPC server. Create a resource-specific server with that
+    // viam RPC server. Add the mock resource to the resource-specific server,
+    // and start the RPC server.
+    std::shared_ptr<sdk::Server> server = std::make_shared<sdk::Server>();
+    ServerType resource_server(server);
+    resource_server.resource_manager()->add(mock->name(), mock);
+    server->start();
+
+    // Create a resource-specific client to the mock over an established
+    // in-process gRPC channel.
+    grpc::ChannelArguments args;
+    std::shared_ptr<grpc::Channel> grpc_channel = TestServer(server).grpc_in_process_channel(args);
```

Comment:
`auto` would be fine here. Also, did you intend for the lifetime of `TestServer` to end on this line? I think it works because you extract a shared pointer. But if you ever wanted `TestServer` to do more (like hold some state) I think this would be fragile. Maybe just make a named `TestServer` object on the stack?

Reply:
> Also, did you intend for the lifetime of TestServer to end on this line?

I did, but I can make an actual object on the stack instead.